subsys/net/l2/ieee802154/ieee802154_radio_csma_ca.c

```diff
@@ -20,44 +23,74 @@ LOG_MODULE_REGISTER(net_ieee802154_csma, CONFIG_NET_L2_IEEE802154_LOG_LEVEL);
 #include "ieee802154_utils.h"
 #include "ieee802154_radio_utils.h"
 
-static inline int csma_ca_radio_send(struct net_if *iface,
-				     struct net_pkt *pkt,
-				     struct net_buf *frag)
+BUILD_ASSERT(CONFIG_NET_L2_IEEE802154_RADIO_CSMA_CA_MIN_BE <=
+		     CONFIG_NET_L2_IEEE802154_RADIO_CSMA_CA_MAX_BE,
+	     "The CSMA/CA min backoff exponent must be less or equal max backoff exponent.");
+
+/* See section 6.2.5.1. */
+static inline int unslotted_csma_ca_radio_send(struct net_if *iface, struct net_pkt *pkt,
+					       struct net_buf *frag)
 {
-	const uint8_t max_bo = CONFIG_NET_L2_IEEE802154_RADIO_CSMA_CA_MAX_BO;
-	const uint8_t max_be = CONFIG_NET_L2_IEEE802154_RADIO_CSMA_CA_MAX_BE;
-	uint8_t retries = CONFIG_NET_L2_IEEE802154_RADIO_TX_RETRIES;
-	struct ieee802154_context *ctx = net_if_l2_data(iface);
-	bool ack_required = prepare_for_ack(ctx, pkt, frag);
 	uint8_t be = CONFIG_NET_L2_IEEE802154_RADIO_CSMA_CA_MIN_BE;
+	struct ieee802154_context *ctx = net_if_l2_data(iface);
+	uint32_t symbol_period, turnaround_time;
+	uint8_t remaining_attempts;
+	bool ack_required;
+	bool is_subg_phy;
 	uint8_t nb = 0U;
-	int ret = -EIO;
+	int ret;
 
 	NET_DBG("frag %p", frag);
 
-loop:
-	while (retries) {
-		retries--;
+	is_subg_phy = ieee802154_get_hw_capabilities(iface) & IEEE802154_HW_SUB_GHZ;
+	/* TODO: Move symbol period calculation to radio driver. */
+	symbol_period = IEEE802154_PHY_SYMBOL_PERIOD(is_subg_phy);
+	turnaround_time = IEEE802154_PHY_A_TURNAROUND_TIME(is_subg_phy);
 
-		if (be) {
-			uint8_t bo_n = sys_rand32_get() & ((1 << be) - 1);
+	/* See section 6.7.4.4 - Retransmissions. */
+	remaining_attempts = CONFIG_NET_L2_IEEE802154_RADIO_TX_RETRIES + 1;
 
-			k_busy_wait(bo_n * 20U);
-		}
+	while (remaining_attempts) {
+		remaining_attempts--;
 
 		while (1) {
-			if (!ieee802154_cca(iface)) {
+			if (be) {
+				uint8_t bo_n = sys_rand32_get() & ((1 << be) - 1);
+
+				/* TODO: k_busy_wait() is too inaccurate on many platforms, the
+				 * radio API should expose a precise radio clock instead (which may
+				 * fall back to k_busy_wait() if the radio does not have a clock).
+				 */
+				k_busy_wait(bo_n * IEEE802154_A_UNIT_BACKOFF_PERIOD_US(
+							   turnaround_time, symbol_period));
+			}
+
+			ret = ieee802154_cca(iface);
+			if (ret == 0) {
+				/* Channel is idle -> CSMA Success */
 				break;
 			}
 
-			be = MIN(be + 1, max_be);
+			/* Channel is not idle -> CSMA Backoff */
+			be = MIN(be + 1, CONFIG_NET_L2_IEEE802154_RADIO_CSMA_CA_MAX_BE);
 			nb++;
 
-			if (nb > max_bo) {
-				goto loop;
+			if (nb > CONFIG_NET_L2_IEEE802154_RADIO_CSMA_CA_MAX_BO) {
+				return -EIO;
 			}
 		}
 
+		ack_required = prepare_for_ack(ctx, pkt, frag);
```

Comment:
A question - wasn't one of the issues we've discussed earlier the fact, that soft CSMA/CA implementation deals with ACK processing and retransmissions, causing problems with HW CSMA/CA support (i. e. no retransmissions then)? I'm having an impression, that we're just cementing this behavior (instead of moving ACK processing out of `aloha_radio_send()`/`unslotted_csma_ca_radio_send()`)?

Reply:
The problem we stumbled upon previously was more on the RX path: Medium access arbitration algorithms have nothing at all to do with handling incoming ACKs. Once I'm on it I can really fix this now.

Nevertheless you're right: Even on the TX path it equally breaks encapsulation to handle ACKs (and therefore retransmissions) in functions that should simply check whether the medium is clear. This should be moved to ieee802154.c->ieee802154_send(). I'll fix that as well.

And thinking more about it now, it seems that checking for IEEE802154_HW_TX_RX_ACK in combination with soft CSMA is also missing, so I'll have to add that as well.

Finally, now that we've agreed that the auto-ack configuration should be deprecated this will also fit well into this PR.

Adding all these additional fixes will take a little more time, though.

Reply:
> @nandojve Thanks for your review! I went through your comments and made changes or explained where I didn't, if you like you can have another look.

@fgrandel thank you for the heads up about `prepare_for_ack`. I'm OK with the changes but it is not clear to me if PR is ready or if you will send more changes.

Reply:
Due to the review comments of rlubos I'll send several additional changes. So this PR is no longer ready.